### PR TITLE
[flang]Add support for -moutline-atomics and -mno-outline-atomics

### PR DESCRIPTION
**Pull request by Leporacanthicus**
This adds the support to add the target-feature.

It will be fully supported once the generic support for adding target-cpu and target-features attributes to functions has landed. See #78289. Regression tests for that will be added once that PR lands.

### CODE REVIEW

File: flang/test/Driver/aarch64-outline-atomics.f90 [hunk 1]
```diff
@@ -0,0 +1,11 @@
+! Test that flang-new forwards the -moutline-atomics and -mno-outline-atomics.
+! RUN: %flang -moutline-atomics --target=aarch64-none-none -### %s -o %t 2>&1  | FileCheck %s
+! CHECK: "-target-feature" "+outline-atomics"
+
+! RUN: %flang -mno-outline-atomics --target=aarch64-none-none -### %s -o %t 2>&1  | FileCheck %s --check-prefix=CHECK-NOOUTLINE
+! CHECK-NOOUTLINE: "-target-feature" "-outline-atomics"
+
+! RUN: %flang -mno-outline-atomics --target=x86-none-none -### %s -o %t 2>&1  | FileCheck %s --check-prefix=CHECK-ERRMSG
+! CHECK-ERRMSG: warning: 'x86' does not support '-mno-outline-atomics'
```
Comment (kiranchandramohan) [inline flang/test/Driver/aarch64-outline-atomics.f90:9]: Should there be a test to check the behaviour when no outline flags are specified and the target is `aarch64`?

File: clang/lib/Driver/ToolChains/Flang.cpp [hunk 2]
```diff
@@ -354,6 +354,27 @@ void Flang::addTargetOptions(const ArgList &Args,
...
+      } else {
+        CmdArgs.push_back("-target-feature");
+        CmdArgs.push_back("-outline-atomics");
+      }
+    }
+  } else if (Triple.isAArch64() &&
+             getToolChain().IsAArch64OutlineAtomicsDefault(Args)) {
+    CmdArgs.push_back("-target-feature");
+    CmdArgs.push_back("+outline-atomics");
+  }
```
Comment (kiranchandramohan) [inline clang/lib/Driver/ToolChains/Flang.cpp:376]: Can this code be shared with `Clang` ?

Comment (banach-space) [reply clang/lib/Driver/ToolChains/Flang.cpp:376]: That's usually done in places like https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/ToolChains/CommonArgs.cpp

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 3]
```diff
@@ -402,6 +402,21 @@ static void parseTargetArgs(TargetOptions &opts, llvm::opt::ArgList &args) {
   for (const llvm::opt::Arg *currentArg :
        args.filtered(clang::driver::options::OPT_target_feature))
     opts.featuresAsWritten.emplace_back(currentArg->getValue());
+
+  llvm::Triple targetTriple{llvm::Triple(opts.triple)};
+  if (const llvm::opt::Arg *A =
+          args.getLastArg(clang::driver::options::OPT_moutline_atomics,
+                          clang::driver::options::OPT_mno_outline_atomics)) {
+    // Option -moutline-atomics supported for AArch64 target only.
+    if (!targetTriple.isAArch64()) {
```
Comment (tblah) [inline flang/lib/Frontend/CompilerInvocation.cpp:411]: Is this condition the wrong way around?

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 4]
```diff
@@ -2790,3 +2790,27 @@ void tools::addHIPRuntimeLibArgs(const ToolChain &TC, Compilation &C,
     }
   }
 }
+
+void tools::addOutlineAtomicsArgs(const Driver &D, const ToolChain &TC, const llvm::opt::ArgList &Args, llvm::opt::ArgStringList & CmdArgs, const llvm::Triple &Triple)
```
Comment (kiranchandramohan) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:2794]: There is probably a clang-format issue here.

File: clang/lib/Driver/ToolChains/CommonArgs.h [hunk 5]
```diff
@@ -215,11 +215,15 @@ void addMachineOutlinerArgs(const Driver &D, const llvm::opt::ArgList &Args,
 void addOpenMPDeviceRTL(const Driver &D, const llvm::opt::ArgList &DriverArgs,
                         llvm::opt::ArgStringList &CC1Args,
                         StringRef BitcodeSuffix, const llvm::Triple &Triple);
+
+void addOutlineAtomicsArgs(const Driver &D, const ToolChain &TC, const llvm::opt::ArgList &Args, llvm::opt::ArgStringList &CmdArgs, const llvm::Triple &Triple);
+
```
Comment (kiranchandramohan) [inline clang/lib/Driver/ToolChains/CommonArgs.h:220]: Nit: clang-format

File: clang/lib/Driver/ToolChains/CommonArgs.h [hunk 6]
```diff
@@ -215,11 +215,15 @@ void addMachineOutlinerArgs(const Driver &D, const llvm::opt::ArgList &Args,
...
 } // end namespace tools
 } // end namespace driver
 } // end namespace clang
 
 clang::CodeGenOptions::FramePointerKind
 getFramePointerKind(const llvm::opt::ArgList &Args, const llvm::Triple &Triple);
 
+
```
Comment (kiranchandramohan) [inline clang/lib/Driver/ToolChains/CommonArgs.h:228]: Nit: accidental change?

File: clang/lib/Driver/ToolChains/Flang.cpp [hunk 7]
```diff
@@ -672,7 +674,7 @@ void Flang::ConstructJob(Compilation &C, const JobAction &JA,
   CmdArgs.push_back(Args.MakeArgString(TripleStr));
 
   if (isa<PreprocessJobAction>(JA)) {
-      CmdArgs.push_back("-E");
+    CmdArgs.push_back("-E");
```
Comment (kiranchandramohan) [inline clang/lib/Driver/ToolChains/Flang.cpp:677]: Nit: accidental change.

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 8]
```diff
@@ -402,6 +402,21 @@ static void parseTargetArgs(TargetOptions &opts, llvm::opt::ArgList &args) {
...
+    // Option -moutline-atomics supported for AArch64 target only.
+    if (targetTriple.isAArch64()) {
+      if (A->getOption().matches(
+              clang::driver::options::OPT_moutline_atomics)) {
+        opts.featuresAsWritten.push_back("+outline-atomics");
+      } else {
+        opts.featuresAsWritten.push_back("-outline-atomics");
+      }
+    }
+  }
```
Comment (kiranchandramohan) [inline flang/lib/Frontend/CompilerInvocation.cpp:419]: Is this required? I did not see a corresponding change in Clang frontend driver.

**Review by kiranchandramohan — APPROVED** [review]
LG. Please address or reply to comments before submitting.

**Comment by clementval** [comment]
This is breaking couple of arm buildbots.

**Comment by clementval** [comment]
If you think this is the right fix then you should push it to solve the buildbots or revert the initial commit.

**Pull request merged**